src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,299 @@
+
+/-- Turn a centroid homomorphism into an additive monoid endomorphism. -/
+def to_End (f : centroid_hom α) : add_monoid.End α := (f : α →+ α)
+
+lemma to_End_injective : injective (centroid_hom.to_End : centroid_hom α → add_monoid.End α) :=
+coe_to_add_monoid_hom_injective
+
+/-- Copy of a `centroid_hom` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def copy (f : centroid_hom α) (f' : α → α) (h : f' = f) :
+  centroid_hom α :=
+{ to_fun := f',
+  map_mul_left' := λ a b, by simp_rw [h, map_mul_left],
+  map_mul_right' := λ a b, by simp_rw [h, map_mul_right],
+  ..f.to_add_monoid_hom.copy f' $ by exact h }
+
+variables (α)
+
+/-- `id` as a `centroid_hom`. -/
+protected def id : centroid_hom α :=
+{ map_mul_left' := λ _ _, rfl,
+  map_mul_right' := λ _ _, rfl,
+  .. add_monoid_hom.id α }
+
+instance : inhabited (centroid_hom α) := ⟨centroid_hom.id α⟩
+
+@[simp] lemma coe_id : ⇑(centroid_hom.id α) = id := rfl
+
+lemma coe_to_add_monoid_hom_id : (centroid_hom.id α : α →+ α) = add_monoid_hom.id α := rfl
+
+variables {α}
+
+@[simp] lemma id_apply (a : α) : centroid_hom.id α a = a := rfl
+
+/-- Composition of `centroid_hom`s as a `centroid_hom`. -/
+def comp (g f : centroid_hom α) : centroid_hom α :=
+{ map_mul_left' := λ a b, (congr_arg g $ f.map_mul_left' _ _).trans $ g.map_mul_left' _ _,
+  map_mul_right' := λ a b, (congr_arg g $ f.map_mul_right' _ _).trans $ g.map_mul_right' _ _,
+  .. g.to_add_monoid_hom.comp f.to_add_monoid_hom }
+
+@[simp] lemma coe_comp (g f : centroid_hom α) : ⇑(g.comp f) = g ∘ f := rfl
+@[simp] lemma comp_apply (g f : centroid_hom α) (a : α) : g.comp f a = g (f a) := rfl
+@[simp] lemma coe_comp_add_monoid_hom (g f : centroid_hom α) :
+  (g.comp f : α →+ α) = (g : α →+ α).comp f := rfl
+@[simp] lemma comp_assoc (h g f : centroid_hom α) : (h.comp g).comp f = h.comp (g.comp f) := rfl
+@[simp] lemma comp_id (f : centroid_hom α) : f.comp (centroid_hom.id α) = f := ext $ λ a, rfl
+@[simp] lemma id_comp (f : centroid_hom α) : (centroid_hom.id α).comp f = f := ext $ λ a, rfl
+
+lemma cancel_right {g₁ g₂ f : centroid_hom α} (hf : surjective f) :
+  g₁.comp f = g₂.comp f ↔ g₁ = g₂ :=
+⟨λ h, ext $ hf.forall.2 $ fun_like.ext_iff.1 h, congr_arg _⟩
+
+lemma cancel_left {g f₁ f₂ : centroid_hom α} (hg : injective g) : g.comp f₁ = g.comp f₂ ↔ f₁ = f₂ :=
+⟨λ h, ext $ λ a, hg $ by rw [←comp_apply, h, comp_apply], congr_arg _⟩
+
+instance : has_zero (centroid_hom α) :=
+⟨{ map_mul_left' := λ a b, (mul_zero _).symm,
+  map_mul_right' := λ a b, (zero_mul _).symm,
+  ..(0 : α →+ α) }⟩
+
+instance : has_one (centroid_hom α) := ⟨centroid_hom.id α⟩
+
+instance : has_add (centroid_hom α) :=
+⟨λ f g,
+  { map_mul_left' := λ a b, by simp [map_mul_left, mul_add],
+    map_mul_right' := λ a b, by simp [map_mul_right, add_mul],
+    ..(f + g : α →+ α) } ⟩
+
+instance : has_mul (centroid_hom α) := ⟨comp⟩
+
+instance has_smul_nat : has_smul ℕ (centroid_hom α) :=
```

Comment:
```suggestion
instance has_nsmul : has_smul ℕ (centroid_hom α) :=
```

Reply:
If I understood @eric-wieser correctly, he wanted the names to match the typeclass name https://github.com/leanprover-community/mathlib/pull/12746#pullrequestreview-1060159820 ?